README.md
# Foreman Exporter for Prometheus

[![Total alerts](https://img.shields.io/lgtm/alerts/g/marcinbojko/foreman_exporter.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/marcinbojko/foreman_exporter/alerts/)
[![Super-Linter](https://github.com/marcinbojko/foreman_exporter/actions/workflows/01_lint_me.yml/badge.svg)](https://github.com/marcinbojko/foreman_exporter/actions/workflows/01_lint_me.yml)

<!-- TOC -->

- [Foreman Exporter for Prometheus](#foreman-exporter-for-prometheus)
  - [What is it](#what-is-it)
    - [Metrics](#metrics)
    - [Foreman presented fields and labels](#foreman-presented-fields-and-labels)
    - [Tested with](#tested-with)
  - [Usage](#usage)
    - [Foreman configuration](#foreman-configuration)
      - [In Foreman, create a role first](#in-foreman-create-a-role-first)
      - [The assign it to an user](#the-assign-it-to-an-user)
    - [Configuration](#configuration)
    - [Docker-compose example](#docker-compose-example)
    - [Prometheus scrape example](#prometheus-scrape-example)
  - [Kubernetes example](#kubernetes-example)
  - [Kubertenes with Service Monitor example](#kubertenes-with-service-monitor-example)
  - [To DO](#to-do)
  - [Tests](#tests)
  - [About](#about)

<!-- /TOC -->

## What is it

Foreman exporter is a small Python3 script exposing host metrics from The Foreman [https://www.theforeman.org/](https://www.theforeman.org/). It uses API to get as much needed information about hosts statuses wrapping these in labels provided by The Foreman itself.
It can be used as a stats tool for configuration applying status, hosts status, and potential errors

### Metrics

- `foreman_exporter_hosts` - Gauge - returns host list with their statuses
- `foreman_exporter_hosts_request_time_seconds` - Gauge - returns time needed to do a  `/api/hosts` query
- `foreman_exporter_hosts_count` - Gauge - total number of hosts returned by query
- `foreman_exporter_dashboard_total_hosts` - Gauge - "Total hosts count"
- `foreman_exporter_dashboard_bad_hosts` - Gauge - "Hosts with error state"
- `foreman_exporter_dashboard_bad_hosts_enabled` - Gauge - "Hosts with error state and alerts enabled"
- `foreman_exporter_dashboard_active_hosts` - Gauge - "Hosts which recently applied changes"
- `foreman_exporter_dashboard_active_hosts_ok` - Gauge - "Hosts which recently applied changes successfully"
- `foreman_exporter_dashboard_active_hosts_ok_enabled` - Gauge - "Hosts which recently applied changes successfully with alerts enabled"
- `foreman_exporter_dashboard_ok_hosts` - Gauge - "Hosts without changes or errors"
- `foreman_exporter_dashboard_ok_hosts_enabled` - Gauge - "Hosts without changes or errors, with alerts enabled"
- `foreman_exporter_dashboard_disabled_hosts` - Gauge - "Hosts with alerts disabled"
- `foreman_exporter_dashboard_pending_hosts` - Gauge - "Hosts that had pending changes"
- `foreman_exporter_dashboard_pending_hosts_enabled` - Gauge - "Hosts that had pending changes with alerts enabled"
- `foreman_exporter_dashboard_out_of_sync_hosts` - Gauge - "Out of sync hosts"
- `foreman_exporter_dashboard_out_of_sync_hosts_enabled` - Gauge - "Out of sync hosts with alerts enabled"
- `foreman_exporter_dashboard_good_hosts` - Gauge - "Hosts without errors"
- `foreman_exporter_dashboard_good_hosts_enabled` - Gauge - "Hosts without errors, with alerts enabled"
- `foreman_exporter_dashboard_percentage` - Gauge - "Hosts without errors percent"
- `foreman_exporter_dashboard_reports_missing` - Gauge - "Hosts which are not reporting"
- `foreman_exporter_dashboard_request_time_seconds` - Gauge - returns time needed to do a `/api/dashboards` query

### Foreman presented fields and labels

|foreman api field|prometheus label|
|-----------------|----------------|
|name|hostname|
|domain_name|domain|
|global_status_label|global_label|
|configuration_status|configuration_status|
|configuration_status_label|configuration_status_label|
|puppet_status|puppet_status|
|environment_name|puppet_environment|
|operatingsystem_name|operatingsystem|
|-|hostname|

### Tested with

- Foreman 2.0.x
- Foreman 2.2.x
- Foreman 1.22.x
- Foreman 1.23.x

## Usage

### Foreman configuration

To use access to API you need an user with at least 'viewer' role for hosts and `access dashboard` role to read dashboards statuses.

#### In Foreman, create a role first

![create_role](./images/01_create_role.png)

#### The assign it to an user

![assign_role](./images/02_assign_role_to_user.png)

### Configuration

Create a set of environment variables. All variables are required.

```ini
FOREMAN_REQUEST_URI=https://foreman.home.lan
FOREMAN_REQUEST_USER=api
FOREMAN_REQUEST_PASSWORD=api
FOREMAN_REQUEST_TLS_VERIFY=false
FOREMAN_REQUEST_TIMEOUT=60
FOREMAN_REQUEST_INTERVAL=120
```

### Docker-compose example

```yaml
version: "3.8"
services:
  foreman_exporter:
    image: marcinbojko/foreman_exporter:latest
    ports:
     - "8000:8000"
    env_file:
      ./foreman_exporter.env
```

### Prometheus scrape example

```yaml
- job_name: 'foreman_exporter'
  scrape_interval: 120s
  honor_labels: true
  metrics_path: '/'
  scheme: http
  static_configs:
    - targets:
      - 'foreman_exporter:8000'
      labels:
        app: "foreman-exporter"
        env: "int"
        team: "it"
```

## Kubernetes example

Working example is available as `./foreman-exporter.yaml`

To use:

- Change ENV Variables in config section of a file:

  ```yaml
  ---
  apiVersion: v1
  kind: ConfigMap
  metadata:
    name: foreman-exporter-env-config
  data:
    FOREMAN_REQUEST_INTERVAL: "120"
    FOREMAN_REQUEST_TIMEOUT: "60"
    FOREMAN_REQUEST_TLS_VERIFY: "false"
    FOREMAN_REQUEST_URI: https://foreman.sample.com
    FOREMAN_REQUEST_USER: api
  ---
  ```

- Generate new secret for `FOREMAN_REQUEST_PASSWORD`

  ```bash
  echo -n newpassword|base64
  ```

  ```yaml
  apiVersion: v1
  kind: Secret
  metadata:
    name: foreman-exporter-env-secret
  data:
    FOREMAN_REQUEST_PASSWORD: base64-password-here-from-above
  ---
  ```

- Change ingress name or/and add tls section

  ```yaml
  ---
   apiVersion: extensions/v1beta1
   kind: Ingress
   metadata:
     name: foreman-exporter-ingress
     labels:
         name: foreman-exporter
   spec:
     rules:
     - host: foreman-exporter.sample.com
       http:
         paths:
         - pathType: Prefix
           path: "/"
           backend:
             serviceName: foreman-exporter-service
             servicePort: 8000
   ```

- Run

  ```bash
  kubectl apply -f foreman-exporter.yaml
  ```

## Kubertenes with Service Monitor example

If you'd like to expose your metrics to Prometheus Operator:

`./k8s/foreman-exporter-custom-metrics.yaml`

## To DO

- add more API checks (facts maybe?)
- ~~Kubernetes setup~~
- improve python skills

## Tests

- pyreq
- isort
- flake8
- pylint
- pylance

## About

That's a small side-project for me to learn Python3


CHANGELOG.md
# Changelog

## 2022-10-26 0.0.14

* fixed `FOREMAN_REQUEST_TLS_VERIFY` handling

## 2022-10-26 0.0.13

* better `json` handling
* switch to `httpx`
* application no longer raises system exception when request fails
* switch to python 3.11
* removed requirements on distutils

## 2021-12-12 0.0.12

* fixed main loop logic
* preparation to logging

## 2021-10-30 0.0.11

* added request for `/api/status` to read foreman's version
* in case of foreman's version higher than `2.2.0` pagination counter set to `all` instead of fake number
* separation of classes to fit `Pylance` and `Pylint` tests
* more verbose logs

## 2021-09-08 0.0.10

* fixed `REQUEST_HOSTS` being empty

## 2021-09-07 0.0.9

* disabled exposing passwords in logs
* image built with latest python
* Dockerfile validation with `hadolint`

## 2021-02-18 0.0.8

* fixed pagination limits (default API and GUI limit) for requests

## 2021-02-14 0.0.7

* reworked k8s example
* added k8s example with service-monitor

## 2021-02-08 0.0.6

* added /api/dashboards to gauges
* added gauges
  * `foreman_exporter_dashboard_total_hosts`
  * `foreman_exporter_dashboard_bad_hosts`
  * `foreman_exporter_dashboard_bad_hosts_enabled`
  * `foreman_exporter_dashboard_active_hosts`
  * `foreman_exporter_dashboard_active_hosts_ok`
  * `foreman_exporter_dashboard_active_hosts_ok_enabled`
  * `foreman_exporter_dashboard_ok_hosts`
  * `foreman_exporter_dashboard_ok_hosts_enabled`
  * `foreman_exporter_dashboard_disabled_hosts`
  * `foreman_exporter_dashboard_pending_hosts`
  * `foreman_exporter_dashboard_pending_hosts_enabled`
  * `foreman_exporter_dashboard_out_of_sync_hosts`
  * `foreman_exporter_dashboard_out_of_sync_hosts_enabled`
  * `foreman_exporter_dashboard_good_hosts`
  * `foreman_exporter_dashboard_good_hosts_enabled`
  * `foreman_exporter_dashboard_percentage`
  * `foreman_exporter_dashboard_reports_missing`
  * `foreman_exporter_dashboard_request_time_seconds`
* rename `environment` label from hosts to `puppet_environment`)
* changed `sleep` main loop into threading

## 2021-02-07 0.0.5

* improved raise exception
* updated README.md
* print formatting

## 2021-02-07 0.0.4

* added Kubernetes usage example
* updated README.md

## 2021-02-06 0.0.3

* fixed clearing `body` variable in loop
* suppressed warnings when connecting to selfsigned page
* enabled stdout in dockerised app
* fixed global=>local variables assignement
* pylint applied
* TZ environment variable added to Dockerfile

## 2021-02-06 0.0.2

* fixed `TypeError: 'NoneType' object is not subscriptable` in `body.total` check
* added healtchecks to Dockerfile

## 2021-02-06 0.0.1

* Initial commit


buildme.sh
#!/bin/bash
release=$(grep "LABEL RELEASE" -i Dockerfile|awk '{print $2}'|cut -d\" -f2)
version=$(grep "LABEL VERSION" -i Dockerfile|awk '{print $2}'|cut -d\" -f2)
echo Version: "$version" found
echo Release: "$release" found
if dockerfilelint Dockerfile; then
  echo "Dockerfilelint passed"
else
  echo "Dockerfilelint errors, correct"
  exit 1
fi
if [ -n "$version" ] && [ -n "$release" ]; then
  docker build --pull --no-cache -t "$release":"$version" .
  build_status=$?
  docker container prune --force
  # let's tag latest
  docker tag "$release":"$version" "$release":latest
else
  echo "No $version found, exiting"
  exit 1
fi
# coverage
if [ "$build_status" == 0 ]; then
  echo "Docker build succeed"
  rm -rfv dive.log||true
  rm -rfv ./.coverage.*||true
  trivy image --output .coverage."$version"_trivy.txt "$release":"$version"
  dive --ci "$release":"$version" > .coverage."$version"_dive.txt
  sudo dockle -f json -o .coverage."$version"_dockle.txt "$release":"$version"
else
 echo "Docker build failed, exiting now"
fi


app/foreman_exporter.py
#!/bin/python3
''' foreman exporter '''
import datetime
import json
import distutils.core
import logging
import os
import sys
import threading
import time
import urllib.parse
import urllib.request
from json.decoder import JSONDecodeError

import httpx
from packaging import version
from prometheus_client import Summary, start_http_server
from prometheus_client.core import REGISTRY, GaugeMetricFamily

# Variables
FOREMAN_HOSTS_BODY = None
FOREMAN_HOSTS_RESPONSE = None
FOREMAN_DASHBOARD_BODY = None
FOREMAN_DASHBOARD_RESPONSE = None
FOREMAN_STATUS_RESPONSE = None
FOREMAN_STATUS_BODY = None
FOREMAN_DASHBOARD_ITEMS = []
FOREMAN_EXPORTER_VERSION = "0.0.14"
FOREMAN_VERSION = None

R_API_DEPTH = "1000"                 # how much elements get with every request to /api/hosts

try:
    # set logging
    logging.basicConfig(format='%(asctime)s %(message)s', encoding='utf-8', level=logging.INFO)
except NameError:
    timestamp = datetime.datetime.now()
    print(timestamp, "Logging failed, returning to defaults")

try:
    # host and uri
    if os.getenv("FOREMAN_REQUEST_URI") is not None:
        REQUEST_URI = str(os.getenv("FOREMAN_REQUEST_URI"))
        if not REQUEST_URI.endswith("/"):
            REQUEST_URI += "/"
    else:
        REQUEST_URI = "https://foreman.sample.com/"
    # display variables
    REQUEST_HOSTNAME = (urllib.parse.urlparse(REQUEST_URI)).netloc
    print(f"REQUEST_URI        = {REQUEST_URI}")
    print(f"REQUEST_HOSTNAME   = {REQUEST_HOSTNAME}")
    # user
    if os.getenv("FOREMAN_REQUEST_USER") is not None:
        REQUEST_USER = str(os.getenv("FOREMAN_REQUEST_USER"))
    else:
        REQUEST_USER = "api"
    print(f"REQUEST_USER       = {REQUEST_USER}")
    # password
    if os.getenv("FOREMAN_REQUEST_PASSWORD") is not None:
        REQUEST_PASSWORD = str(os.getenv("FOREMAN_REQUEST_PASSWORD"))
    else:
        REQUEST_PASSWORD = "api"
    # tls_verify
    if os.getenv("FOREMAN_REQUEST_TLS_VERIFY") is not None:
        REQUEST_TLS_VERIFY = bool(distutils.util.strtobool((os.getenv("FOREMAN_REQUEST_TLS_VERIFY"))))
    else:
        REQUEST_TLS_VERIFY = False
    print(f"REQUEST_TLS_VERIFY = {REQUEST_TLS_VERIFY}")
    # request timeout
    if os.getenv("FOREMAN_REQUEST_TIMEOUT") is not None:
        REQUEST_TIMEOUT = int(os.getenv("FOREMAN_REQUEST_TIMEOUT"))
    else:
        REQUEST_TIMEOUT = 15
    print(f"REQUEST_TIMEOUT    = {REQUEST_TIMEOUT}")
    # request interval
    if os.getenv("FOREMAN_REQUEST_INTERVAL") is not None:
        REQUEST_INTERVAL = int(os.getenv("FOREMAN_REQUEST_INTERVAL"))
    else:
        REQUEST_INTERVAL = 30
    print(f"REQUEST_INTERVAL   = {REQUEST_INTERVAL}")
except NameError:
    timestamp = datetime.datetime.now()
    print(timestamp, "Evaluation of Environmental Variables failed, returning to defaults")

if all(v_check is not None for v_check in [REQUEST_URI, REQUEST_HOSTNAME, REQUEST_USER, REQUEST_PASSWORD, REQUEST_TLS_VERIFY, REQUEST_TIMEOUT, REQUEST_INTERVAL]):
    pass
else:
    timestamp = datetime.datetime.now()
    print(timestamp, "One of variables is empty")
    raise SystemExit(1)


# initial data
def f_requests_status():
    ''' Process Foreman's status response '''
    global FOREMAN_STATUS_RESPONSE
    global FOREMAN_STATUS_BODY
    global R_API_DEPTH
    try:
        response = httpx.get(REQUEST_URI+'api/status', auth=(REQUEST_USER, REQUEST_PASSWORD), verify=REQUEST_TLS_VERIFY, timeout=REQUEST_TIMEOUT)
        response.raise_for_status()
        try:
            body = json.loads(response.text)
        except JSONDecodeError:
            print(datetime.datetime.now(), f"Request at: {REQUEST_URI}api/status failed, response is not json")
            time.sleep(1)
            sys.exit()
        if 200 >= response.status_code <= 399:
            print(datetime.datetime.now(), f"Status request at {REQUEST_URI}api/status with code {response.status_code} took {response.elapsed.seconds} seconds")
            FOREMAN_STATUS_BODY = body
            FOREMAN_STATUS_RESPONSE = response
            # Status section
            if FOREMAN_STATUS_BODY['version'] is not None:
                r_status_version = str(FOREMAN_STATUS_BODY['version'])
                print(datetime.datetime.now(), f"Foreman version found at: {REQUEST_URI} is {r_status_version}")
                # since version 2.2.0 pagination can be set to all or number
                if version.parse(r_status_version) > version.parse("2.2.0"):
                    R_API_DEPTH = "all"
            else:
                pass
        else:
            print(datetime.datetime.now(), f"Response code not proper: {response.status_code}")
    except httpx.HTTPStatusError as err:
        print(datetime.datetime.now(), f"Request at: {REQUEST_URI}api/status failed with code {err}")
        time.sleep(1)


def f_requests_hosts():
    ''' Process Foreman's hosts response '''
    global FOREMAN_HOSTS_RESPONSE
    global FOREMAN_HOSTS_BODY
    try:
        response = httpx.get(REQUEST_URI+'api/hosts?per_page='+str(R_API_DEPTH), auth=(REQUEST_USER, REQUEST_PASSWORD), verify=REQUEST_TLS_VERIFY, timeout=REQUEST_TIMEOUT)
        response.raise_for_status()
        try:
            body = json.loads(response.text)
        except JSONDecodeError:
            print(datetime.datetime.now(), f"Request at: {REQUEST_URI}api/hosts failed, response is not json")
            time.sleep(1)
            sys.exit()
        if 200 >= response.status_code <= 399:
            print(datetime.datetime.now(), f"Hosts request at {REQUEST_URI}api/hosts with code {response.status_code} took {response.elapsed.seconds} seconds")
            FOREMAN_HOSTS_BODY = body
            FOREMAN_HOSTS_RESPONSE = response
        else:
            print(datetime.datetime.now(), f"Response code not proper: {response.status_code}")
    except httpx.HTTPStatusError as err:
        print(datetime.datetime.now(), f"Request at: {REQUEST_URI}api/hosts failed with code {err}")
        time.sleep(1)


def f_requests_dashboard():
    ''' Process Foreman's dashboard response '''
    global FOREMAN_DASHBOARD_RESPONSE
    global FOREMAN_DASHBOARD_BODY
    try:
        response = httpx.get(REQUEST_URI+'api/dashboard?per_page=10000', auth=(REQUEST_USER, REQUEST_PASSWORD), verify=REQUEST_TLS_VERIFY, timeout=REQUEST_TIMEOUT)
        response.raise_for_status()
        try:
            body = json.loads(response.text)
        except JSONDecodeError:
            print(datetime.datetime.now(), f"Request at: {REQUEST_URI}api/dashboard failed, response is not json")
            time.sleep(1)
            sys.exit()
        if 200 >= response.status_code <= 399:
            print(datetime.datetime.now(), f"Dashboard request at: {REQUEST_URI}api/dashboard with code {response.status_code} took {response.elapsed.seconds} seconds")
            FOREMAN_DASHBOARD_BODY = body
            FOREMAN_DASHBOARD_RESPONSE = response
        else:
            print(datetime.datetime.now(), f"Response code not proper:  {response.status_code}")
    except httpx.HTTPStatusError as err:
        print(datetime.datetime.now(), f"Request at: {REQUEST_URI}api/dashboard failed with code {err}")
        time.sleep(1)


# register class
class RequestsHosts:
    ''' Register Prometheus Metrics for Foremant's hosts '''
    def __init__(self):
        pass

    @staticmethod
    def collect():
        ''' Register Prometheus Metrics for Foremant's hosts '''
        # foremans gauges
        g_hosts = GaugeMetricFamily("foreman_exporter_hosts", 'foreman host status', labels=['hostname', 'domain', 'configuration', 'configuration_label',
                                    'puppet_status', 'global_label', 'puppet_environment', 'operatingsystem', 'foreman_hostname'])
        if FOREMAN_HOSTS_BODY is not None:
            for each in FOREMAN_HOSTS_BODY['results']:
                name = str(each['name'])
                domain = str(each['domain_name'])
                status = (each['global_status'])
                global_label = str(each['global_status_label'])
                configuration_status = str(each['configuration_status'])
                configuration_status_label = str(each['configuration_status_label'])
                puppet_status = str(each['puppet_status'])
                environment_name = str(each['environment_name'])
                operatingsystem = str(each['operatingsystem_name'])
                if (
                  name is None or domain is None or status is None or configuration_status is None or configuration_status_label is None
                  ):
                    continue
                if global_label is None:
                    global_label = '199'
                if puppet_status is None:
                    puppet_status = '199'
                if environment_name is None:
                    environment_name = 'unknown'
                if operatingsystem is None:
                    operatingsystem = 'unknown'
                g_hosts.add_metric([name, domain, configuration_status, configuration_status_label, puppet_status, global_label,
                                    environment_name, operatingsystem, REQUEST_HOSTNAME], status)
            yield g_hosts
        else:
            pass
        # How long the process was made
        if FOREMAN_HOSTS_RESPONSE is not None:
            g_hosts_time = GaugeMetricFamily("foreman_exporter_hosts_request_time_seconds", 'foreman host request time seconds', labels=['foreman_hostname'])
            g_hosts_time.add_metric([REQUEST_HOSTNAME], int(FOREMAN_HOSTS_RESPONSE.elapsed.seconds))
            yield g_hosts_time
        else:
            pass
        # number of hosts
        if FOREMAN_HOSTS_BODY is not None:
            g_hosts_count = GaugeMetricFamily("foreman_exporter_hosts_count", 'foreman host count', labels=['foreman_hostname'])
            g_hosts_count.add_metric([REQUEST_HOSTNAME], int(FOREMAN_HOSTS_BODY['total']))
            yield g_hosts_count
        else:
            pass


class RequestsDashboard:
    ''' Register Prometheus Metrics for Foremant's Dashboards '''
    def __init__(self):
        pass

    @staticmethod
    def collect():
        ''' Register Prometheus Metrics for Foremant's Dashboards '''
        # Dashboard section
        g_dashboard = {}
        FOREMAN_DASHBOARD_ITEMS.clear()
        if FOREMAN_DASHBOARD_BODY is not None:
            for each in FOREMAN_DASHBOARD_BODY:
                FOREMAN_DASHBOARD_ITEMS.append(each)
                # total hosts
            if FOREMAN_DASHBOARD_ITEMS is not None:
                for each in FOREMAN_DASHBOARD_ITEMS:
                    if each != 'glossary':
                        g_dashboard[each] = GaugeMetricFamily("foreman_exporter_dashboard_"+each, FOREMAN_DASHBOARD_BODY['glossary'][each], labels=['foreman_hostname'])
                        g_dashboard[each].add_metric([REQUEST_HOSTNAME], int(FOREMAN_DASHBOARD_BODY[each]))
                        yield g_dashboard[each]
        else:
            pass
        # How long the process (request dashboard) was made
        if FOREMAN_DASHBOARD_RESPONSE is not None:
            g_dashboard_time = GaugeMetricFamily("foreman_exporter_dashboard_request_time_seconds", 'foreman dashboard request time seconds', labels=['foreman_hostname'])
            g_dashboard_time.add_metric([REQUEST_HOSTNAME], int(FOREMAN_DASHBOARD_RESPONSE.elapsed.seconds))
            yield g_dashboard_time
        else:
            pass


class RequestsStatus:
    ''' Register Prometheus Metrics for Foreman's Status '''
    def __init__(self):
        pass

    @staticmethod
    def collect():
        ''' Register Prometheus Metrics for Foremant's Status '''
#        global R_API_DEPTH
        # How long the process (request dashboard) was made
        if FOREMAN_STATUS_RESPONSE is not None:
            g_status_time = GaugeMetricFamily("foreman_exporter_status_request_time_seconds", 'foreman status request time seconds', labels=['foreman_hostname'])
            g_status_time.add_metric([REQUEST_HOSTNAME], int(FOREMAN_STATUS_RESPONSE.elapsed.seconds))
            yield g_status_time
        else:
            pass


# Create a metric to track time spent and requests made.
REQUEST_TIME = Summary('request_processing_seconds', 'Time spent processing request')


@REQUEST_TIME.time()
def f_process_request():
    """A dummy function that takes some time."""
    time.sleep(REQUEST_INTERVAL)


def f_start_http():
    ''' Start http server '''
    print(datetime.datetime.now(), "Starting http server at port 8000")
    # Start up the server to expose the metrics.
    start_http_server(8000)


def main():
    ''' Main threading loop '''
    thread = threading.Thread(target=f_process_request)
    thread.start()
    thread.join()
    f_requests_status()
    f_requests_hosts()
    f_requests_dashboard()
    f_process_request()


if __name__ == '__main__':
    # Initial fill in
    f_requests_status()
    f_requests_hosts()
    f_requests_dashboard()
    print(datetime.datetime.now(), f"Script version is {FOREMAN_EXPORTER_VERSION}")
    f_start_http()
    # Register gauges
    REGISTRY.register(RequestsHosts())
    REGISTRY.register(RequestsDashboard())
    REGISTRY.register(RequestsStatus())
    while True:
        main()
